src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+            Some(db_len)
+        } else {
+            // Double buffer mode must not be enabled if we haven't been given a second buffer
+            if config.double_buffer {
+                return Err(DMAError::NoSecondBuffer);
+            }
+            None
+        };
+
+        let n_transfers = if let Some(db) = db_len {
+            buf_len.min(db) as u16
+        } else {
+            buf_len as u16
+        };
+        stream.set_number_of_transfers(n_transfers);
+
+        let mut transfer = Self {
+            stream,
+            _channel: PhantomData,
+            peripheral,
+            _direction: PhantomData,
+            buf: Some(memory),
+            double_buf,
+            transfer_length: n_transfers,
+        };
+        transfer.apply_config(config);
+
+        Ok(transfer)
+    }
+
+    /// Starts the transfer, the closure will be executed right after enabling the stream.
+    pub fn start<F>(&mut self, f: F)
+    where
+        F: FnOnce(&mut PERIPHERAL),
+    {
+        // "Preceding reads and writes cannot be moved past subsequent writes"
+        compiler_fence(Ordering::Release);
+
+        unsafe {
+            self.stream.enable();
+        }
+        f(&mut self.peripheral);
+    }
+
+    /// Pauses the dma stream, the closure will be executed right before disabling the stream.
+    pub fn pause<F>(&mut self, f: F)
+    where
+        F: FnOnce(&mut PERIPHERAL),
+    {
+        f(&mut self.peripheral);
+        self.stream.disable()
+    }
+
+    /// Changes the buffer and restarts or continues a double buffer transfer. This must be called
+    /// immediately after a transfer complete event. Returns the old buffer together with its
+    /// `CurrentBuffer`, if an error occurs, this methods will return the new buffer with the error.
+    pub fn next_transfer(
+        &mut self,
+        mut new_buf: BUF,
+    ) -> Result<(BUF, CurrentBuffer), DMAError<BUF>> {
+        if !STREAM::get_transfer_complete_flag() {
+            return Err(DMAError::NotReady(new_buf));
+        }
+        self.stream.clear_transfer_complete_interrupt();
+
+        if self.double_buf.is_some() && DIR::direction() != DmaDirection::MemoryToMemory {
+            // NOTE(unsafe) We now own this buffer and we won't call any &mut methods on `new` until
+            // the end of the DMA transfer
+            let (new_buf_ptr, new_buf_len) = unsafe { new_buf.write_buffer() };
+
+            // We can't change the transfer length while double buffering
+            if new_buf_len < usize::from(self.transfer_length) {
+                return Err(DMAError::SmallBuffer(new_buf));
+            }
+
+            if STREAM::current_buffer() == CurrentBuffer::DoubleBuffer {
+                self.stream.set_memory_address(new_buf_ptr as u32);
+                let old_buf = self.buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
```

Comment:
`self.buf.replace` does a read, but you are right, the replace and all buffer operations could still be moved to before the address setting, will fix this.

Reply:
You can't really rely on `self.buf.replace` because it doesn't do a volatile read, right? So it might be optimized away.

Reply:
I don't think it would in this case since the return result depends on that, I also have the impression that the compiler fence would pass before that, since it usually states about program order. However, like I said I'm not sure if it would prevent reordering to before the register write, so I will go with `SeqCst`.